tornado/iostream.py

```diff
@@ -548,26 +552,26 @@ def _handle_events(self, fd, events):
                 self.io_loop.update_handler(self.fileno(), self._state)
         except UnsatisfiableReadError as e:
             gen_log.info("Unsatisfiable read, closing connection: %s" % e)
-            self.close(exc_info=True)
+            self.close(exc_info=e)
         except Exception:
             gen_log.error("Uncaught exception, closing connection.",
                           exc_info=True)
-            self.close(exc_info=True)
+            self.close(exc_info=e)
             raise
 
     def _run_callback(self, callback, *args):
         def wrapper():
             self._pending_callbacks -= 1
             try:
                 return callback(*args)
-            except Exception:
+            except Exception as e:
                 app_log.error("Uncaught exception, closing connection.",
                               exc_info=True)
                 # Close the socket on an uncaught exception from a user callback
                 # (It would eventually get closed when the socket object is
                 # gc'd, but we don't want to rely on gc happening before we
                 # run out of file descriptors)
-                self.close(exc_info=True)
+                self.close(exc_info=e)
```

Comment:
This loses stack trace information on python 2. That's true throughout this commit, but in most cases the information being lost is not that interesting because the `try` block has a very narrow scope. This one wraps a user-supplied callback, though, so the stack information might be useful. (we can also see the difference because the log line above uses `exc_info=True` while none of the other logs care enough about the stack trace to use it). I think this one should continue to use `exc_info=True` or explicitly capture `sys.exc_info()` at the start of the `except` block.